lib/block-supports/layout.php

```diff
@@ -316,13 +316,60 @@ function gutenberg_get_classnames_from_last_tag( $html ) {
  * @return string                Filtered block content.
  */
 function gutenberg_render_layout_support_flag( $block_content, $block ) {
-	$block_type     = WP_Block_Type_Registry::get_instance()->get_registered( $block['blockName'] );
-	$support_layout = block_has_support( $block_type, array( '__experimentalLayout' ), false );
+	$block_type       = WP_Block_Type_Registry::get_instance()->get_registered( $block['blockName'] );
+	$support_layout   = block_has_support( $block_type, array( '__experimentalLayout' ), false );
+	$has_child_layout = isset( $block['attrs']['style']['layout']['selfStretch'] );
 
-	if ( ! $support_layout ) {
+	if ( ! $support_layout
+	&& ! $has_child_layout ) {
 		return $block_content;
 	}
 
+	$outer_class_names = array();
+
+	if ( $has_child_layout && ( 'fixed' === $block['attrs']['style']['layout']['selfStretch'] || 'fill' === $block['attrs']['style']['layout']['selfStretch'] ) ) {
+
+		$container_content_class = wp_unique_id( 'wp-container-content-' );
+
+		$child_layout_styles = array();
+
+		if ( 'fixed' === $block['attrs']['style']['layout']['selfStretch'] && isset( $block['attrs']['style']['layout']['flexSize'] ) ) {
+			$child_layout_styles[] = array(
+				'selector'     => ".$container_content_class",
+				'declarations' => array(
+					'flex-shrink' => '0',
```

Comment:
Oh, this is an interesting one. I assume this is to attempt to preserve what a user has chosen?

Will we always want to prevent the element from shrinking? That is, if someone sets a large value for flex-basis (e.g. `900px`) I was wondering if we'll want to preserve that in smaller viewports where it'll cause an overflow, or in desktop viewports where it'll also overflow `contentSize`?

Reply:
Yes, if we set a fixed size we do want to prevent it from shrinking. That's touched upon in [this comment](https://github.com/WordPress/gutenberg/issues/44467#issuecomment-1261911573), but also if we _don't_ set `flex-shrink` there's no way to make sure the block actually sticks to the defined size 😅 
As users are deliberately setting values, I think it's fine for them to overflow their containers where the values are too large.

Reply:
> As users are deliberately setting values, I think it's fine for them to overflow their containers where the values are too large.

Ah, I see your point! It's a tricky one to work out the desired behaviour here (whether to honour what the user is attempting to do, or try to make sure they don't break the layout). I think I probably lean more toward preventing them from overflowing the container, because otherwise it breaks outside of the content size / wide size logic, which could be unexpected, and any width that's set greater than `400px` would overflow most phone's viewports in a vertical orientation. Here's a paragraph set to `500px` on desktop and mobile:

| Desktop | Mobile |
| --- | --- |
| <img width="1067" alt="image" src="https://user-images.githubusercontent.com/14988353/199648182-44791ea6-4aba-46da-b02b-725a9d6667ec.png"> | <img width="478" alt="image" src="https://user-images.githubusercontent.com/14988353/199648207-578054ea-b1e7-4e0f-9720-46b98078f968.png"> |

From a quick look at the Column block's ad hoc width control, it appears that it currently doesn't include a `flex-shrink: 0` rule, so it currently ensures that Columns blocks don't break out.

Either way, though, since these CSS rules aren't baked into post content, it looks like it'll be a simple rule to change if we want to 🙂

Reply:
Interesting balance to strike on this one. In general, it seems like we try to start from a position of encouraging good design choices by default, then move towards honouring user choices. Following that, I'd lean a little more toward not allowing the broken layout first, then honouring explicit choices second.

That's only my naive two cents though, so take it with a grain of salt 🤷‍♂️

Reply:
I'm not sure it's helpful to think of this situation as a "broken layout", because people might want to create a horizontally scrolling page, and currently there's no way to do so in the site editor. It's easy enough to undo a change, and obvious enough that a horizontal scroll is happening if the user _does_ add a bunch of fixed widths by accident, that we should let it happen. 

Also, as I said before, there's no way of making this feature work properly _without_ the flex-shrink in place. If we show a control to fix the width of a block, and that control doesn't actually fix it, that's a much more broken experience than allowing users the possibility of making their page scroll horizontally.

Reply:
I think it's a really interesting argument in both directions here (whether to honour the user's explicit choice, or attempt to do a fuzzy approximation). With other width controls (e.g. Columns or Search block), it looks like we currently don't allow overflowing the container? The search block sets the `width` value explicitly, but also has `max-width: 100%` so it doesn't overflow.

In terms of which user intent to prioritise, I'd be curious to hear thoughts from @jasmussen on this one — for a bit of history, there's been prior attempts to land a width block support, and the responsive behaviour was flagged on an earlier attempt here: https://github.com/WordPress/gutenberg/pull/32502#issuecomment-863787000, and there's a subsequent open issue for responsive / intrinsic web design and how it relates to block controls in: https://github.com/WordPress/gutenberg/issues/34641.

Whichever way we prioritise, I think it'd be good to figure out how we can support the one we haven't prioritised. E.g. if we went with the fuzzy approach, how do we then allow folks to intentionally create horizontally scrolling layouts, and alternately, if we honour the explicit pixel width, how do we allow folks to set a larger size that then shrinks in smaller viewports.

Reply:
It was @jasmussen's recommendation [here](https://github.com/WordPress/gutenberg/issues/44467#issuecomment-1261911573) to let explicit widths break out of their container 😄

Reply:
Ah, good catch, thanks @tellthemachines, I missed that one! That's a more recent comment so more likely captures the current thinking 🙂

Reply:
No worries, it's always good to poke at stuff anyway and see how we can improve it!

I think in this case the fuzzy approach would add too much complexity for the value it might bring: removing the flex-shrink would stop the setting from working as expected, so we'd probably have to do some calculation of parent widths or something horrible like that 😬  

Also, given we already have Columns with the width constraints, it makes sense to have this behave differently, so folks can use Columns for a content-width-respecting layout and Row for more freestyle options.

I'm actually pretty excited to finally have a way to create fixed width, horizontally scrolling pages tbh 😁

Reply:
Great points, and there's nothing stopping us from adding a `fluid` or fuzzy option as an additional control alongside `fixed` in a follow-up at some point to handle the case where someone wants to set a target width that gets either shrunk or clampified 👍